corehq/apps/es/README.rst

```diff
@@ -0,0 +1,220 @@
+- allows HQ Elasticsearch index state to remain decoupled from the
+  commcare-cloud codebase
+- is not disruptive -- does not prohibit any other kind of standard maintenance
+  that might come up while the operation is underway
+- is "fire and forget" -- does not require active polling of intermediate state
+  in order to progress the overall operation
+- is practical for third party HQ hosters to use
+
+One way to accomplish these constraints is to implement an "index multiplexing"
+feature in HQ, where Elasticsearch write operations are duplicated across two
+indexes. This design facilitates maintaining two up-to-date versions of any
+index (a primary read/write index and a secondary write-only index),
+allowing HQ to run in a "normal" state (i.e. not a custom "maintenance" state)
+while providing the ability to switch back and forth (swapping primary and
+secondary) before fully committing to abandoning one of them. Creating a copy of
+an index is the unavoidable nature of a reindex operation, and multiplexing
+allows safe switching from one to the other without causing disruptions or
+outages while keeping both up-to-date.
+
+The least disruptive way to accomplish a multiplexing design is with an adapter
+layer that operates between the low-level third party Elasticsearch Python
+client library and high-level HQ components which need to read/write data in an
+Elasticsearch index. Conveniently, HQ already has the initial framework for this
+layer (the ``ElasticsearchInterface`` class), so the adapter layer is not a new
+concept. The reason the ``ElasticsearchInterface`` implementation cannot
+accommodate multiplexing is because it is not widely adopted for low-level
+Elasticsearch operations (e.g. index creation and configuration), and it does
+not sufficiently decouple low-level Elasticsearch document manipulation logic
+from the high-level HQ code that uses it.
+
+With a multiplexing adapter layer, reindexing an Elasticsearch index can be
+as few as four concise steps, none of which are time-critical in respect to each
+other:
+
+1. Merge and deploy a PR that configures multiplexing on an index.
+2. Execute an idempotent management command that updates the secondary index
+   from its primary counterpart.
+3. Merge and deploy a PR that disables multiplexing for the index, (now using
+   only the new index).
+4. Execute a management command to delete the old index.
+
+**Note**: the above steps are not limited to a single index at a time. That is,
+the implementation does not prohibit configuring multiplexing and reindexing
+multiple indexes at once.
+
+This reindex procedure is inherently safe because:
+
+- At any point in the process, the rollback procedure is a simple code change
+  (i.e. revert PR, deploy).
+- The operation responsible for populating the secondary index is idempotent
+  *and* decoupled from the index configuration, allowing it to undergo change
+  iterations without aborting the entire process (thereby losing reindex
```

Comment:
Ideally there are none, but the model supports it. If the "sync and verify" tool comes back saying the indexes are not equivalent, it can be changed (fixed) and run again without urgency or discarding of existing sync progress.

Reply:
>  "sync and verify" tool 

What tool is this?

Reply:
I was referring to the management command outlined in step **2.** of the reindexing procedure in the README doc:

```markdown
2. Execute an idempotent management command that updates the secondary index
   from its primary counterpart.
```